File: tools/project/src/project/cache.py
"""Cache management for project index."""

import hashlib
import json
import time
from pathlib import Path
from typing import Optional

from .database import ProjectDatabase


def get_cache_dir() -> Path:
    """Get the cache directory."""
    xdg_cache = Path.home() / ".cache"
    if "XDG_CACHE_HOME" in __import__("os").environ:
        xdg_cache = Path(__import__("os").environ["XDG_CACHE_HOME"])
    return xdg_cache / "daedalos" / "project"


class CacheManager:
    """Manages the cache for a project."""

    def __init__(self, project_path: Path):
        self.project_path = project_path.resolve()
        self.cache_dir = self._get_cache_path()
        self.cache_dir.mkdir(parents=True, exist_ok=True)
        self.meta_file = self.cache_dir / "meta.json"
        self.db_file = self.cache_dir / "index.db"
        self._db: Optional[ProjectDatabase] = None

    def _get_cache_path(self) -> Path:
        """Get cache path for this project."""
        # Hash the project path to create unique cache directory
        path_hash = hashlib.sha256(str(self.project_path).encode()).hexdigest()[:16]
        return get_cache_dir() / path_hash

    def get_database(self) -> ProjectDatabase:
        """Get or create the database."""
        if self._db is None:
            self._db = ProjectDatabase(self.db_file)
        return self._db

    def close(self):
        """Close database connection."""
        if self._db:
            self._db.close()
            self._db = None

    def is_stale(self, max_age_seconds: int = 3600) -> bool:
        """Check if cache is stale."""
        if not self.meta_file.exists():
            return True

        try:
            meta = json.loads(self.meta_file.read_text())
            last_indexed = meta.get("last_indexed", 0)
            return time.time() - last_indexed > max_age_seconds
        except (json.JSONDecodeError, KeyError):
            return True

    def mark_fresh(self):
        """Mark cache as freshly indexed."""
        meta = self._load_meta()
        meta["last_indexed"] = time.time()
        meta["project_path"] = str(self.project_path)
        self._save_meta(meta)

    def get_last_indexed(self) -> Optional[float]:
        """Get timestamp of last indexing."""
        meta = self._load_meta()
        return meta.get("last_indexed")

    def set_project_type(self, project_type: str):
        """Store detected project type."""
        meta = self._load_meta()
        meta["project_type"] = project_type
        self._save_meta(meta)

    def get_project_type(self) -> Optional[str]:
        """Get stored project type."""
        meta = self._load_meta()
        return meta.get("project_type")

    def _load_meta(self) -> dict:
        """Load metadata from file."""
        if self.meta_file.exists():
            try:
                return json.loads(self.meta_file.read_text())
            except json.JSONDecodeError:
                return {}
        return {}

    def _save_meta(self, meta: dict):
        """Save metadata to file."""
        self.meta_file.write_text(json.dumps(meta, indent=2))

    def clear(self):
        """Clear all cached data."""
        if self._db:
            self._db.close()
            self._db = None

        if self.db_file.exists():
            self.db_file.unlink()
        if self.meta_file.exists():
            self.meta_file.unlink()

File: crates/project/src/cache.rs
//! Cache management for project index

use std::path::{Path, PathBuf};
use std::time::{SystemTime, UNIX_EPOCH};

use anyhow::{Context, Result};
use serde::{Deserialize, Serialize};
use sha2::{Digest, Sha256};

use crate::database::ProjectDatabase;

/// Cache metadata
#[derive(Debug, Clone, Serialize, Deserialize, Default)]
struct CacheMeta {
    last_indexed: f64,
    project_path: String,
    project_type: Option<String>,
}

/// Cache manager for a project
pub struct CacheManager {
    project_path: PathBuf,
    cache_dir: PathBuf,
    meta_file: PathBuf,
    db_file: PathBuf,
    db: Option<ProjectDatabase>,
}

impl CacheManager {
    /// Create a new cache manager for a project
    pub fn new(project_path: &Path) -> Result<Self> {
        let project_path = project_path.to_path_buf();
        let cache_dir = get_cache_path(&project_path)?;

        std::fs::create_dir_all(&cache_dir).context("Failed to create cache directory")?;

        let meta_file = cache_dir.join("meta.json");
        let db_file = cache_dir.join("index.db");

        Ok(Self {
            project_path,
            cache_dir,
            meta_file,
            db_file,
            db: None,
        })
    }

    /// Get or create the database
    pub fn get_database(&mut self) -> Result<&ProjectDatabase> {
        if self.db.is_none() {
            self.db = Some(ProjectDatabase::new(&self.db_file)?);
        }
        Ok(self.db.as_ref().unwrap())
    }

    /// Open a new database connection (for use when you need separate ownership)
    pub fn open_database(&self) -> Result<ProjectDatabase> {
        ProjectDatabase::new(&self.db_file)
    }

    /// Get the database file path
    pub fn db_path(&self) -> &Path {
        &self.db_file
    }

    /// Check if the cache is stale (older than max_age_seconds)
    pub fn is_stale(&self) -> bool {
        self.is_stale_with_age(3600) // 1 hour default
    }

    /// Check if the cache is stale with custom max age
    pub fn is_stale_with_age(&self, max_age_seconds: u64) -> bool {
        if !self.meta_file.exists() {
            return true;
        }

        match self.load_meta() {
            Ok(meta) => {
                let now = SystemTime::now()
                    .duration_since(UNIX_EPOCH)
                    .map(|d| d.as_secs_f64())
                    .unwrap_or(0.0);
                now - meta.last_indexed > max_age_seconds as f64
            }
            Err(_) => true,
        }
    }

    /// Mark the cache as freshly indexed
    pub fn mark_fresh(&mut self) -> Result<()> {
        let mut meta = self.load_meta().unwrap_or_default();
        meta.last_indexed = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map(|d| d.as_secs_f64())
            .unwrap_or(0.0);
        meta.project_path = self.project_path.to_string_lossy().to_string();
        self.save_meta(&meta)
    }

    /// Get the stored project type
    pub fn get_project_type(&self) -> Option<String> {
        self.load_meta().ok().and_then(|m| m.project_type)
    }

    /// Set the project type
    pub fn set_project_type(&mut self, project_type: &str) -> Result<()> {
        let mut meta = self.load_meta().unwrap_or_default();
        meta.project_type = Some(project_type.to_string());
        self.save_meta(&meta)
    }

    /// Clear all cached data
    pub fn clear(&mut self) -> Result<()> {
        self.db = None;

        if self.db_file.exists() {
            std::fs::remove_file(&self.db_file)?;
        }
        if self.meta_file.exists() {
            std::fs::remove_file(&self.meta_file)?;
        }
        Ok(())
    }

    fn load_meta(&self) -> Result<CacheMeta> {
        let content = std::fs::read_to_string(&self.meta_file)?;
        let meta: CacheMeta = serde_json::from_str(&content)?;
        Ok(meta)
    }

    fn save_meta(&self, meta: &CacheMeta) -> Result<()> {
        let content = serde_json::to_string_pretty(meta)?;
        std::fs::write(&self.meta_file, content)?;
        Ok(())
    }
}

/// Get the cache directory for Daedalos
fn get_cache_dir() -> Result<PathBuf> {
    let cache_dir = if let Ok(xdg_cache) = std::env::var("XDG_CACHE_HOME") {
        PathBuf::from(xdg_cache)
    } else if let Some(home) = dirs::home_dir() {
        home.join(".cache")
    } else {
        return Err(anyhow::anyhow!("Could not determine cache directory"));
    };

    Ok(cache_dir.join("daedalos").join("project"))
}

/// Get the cache path for a specific project
fn get_cache_path(project_path: &Path) -> Result<PathBuf> {
    let base = get_cache_dir()?;
    let path_str = project_path.to_string_lossy();

    let mut hasher = Sha256::new();
    hasher.update(path_str.as_bytes());
    let hash = hex::encode(&hasher.finalize()[..8]);

    Ok(base.join(hash))
}
